File: articles/r-chemistry-statistics-experimental-analysis/fortran/r_chemistry_kernel.f90
program r_chemistry_kernel
  implicit none

  real(8), dimension(3) :: x
  real(8) :: mean_x, sd_x, se_x, rsd, unknown
  integer :: n

  x = (/1.02d0, 1.05d0, 0.99d0/)
  n = 3

  mean_x = sum(x) / n
  sd_x = sqrt(sum((x - mean_x) ** 2) / (n - 1))
  se_x = sd_x / sqrt(real(n, 8))
  rsd = 100.0d0 * sd_x / mean_x
  unknown = (0.95d0 - 0.02d0) / 0.30d0

  print *, "mean=", mean_x
  print *, "sample_sd=", sd_x
  print *, "standard_error=", se_x
  print *, "rsd_percent=", rsd
  print *, "unknown_concentration=", unknown

end program r_chemistry_kernel

File: articles/r-chemistry-statistics-experimental-analysis/julia/r_chemistry_kernel.jl
# R chemistry numerical kernel in Julia.
# Mirrors simple statistical chemistry calculations.

function mean_value(x)
    return sum(x) / length(x)
end

function sample_sd(x)
    xbar = mean_value(x)
    return sqrt(sum((x .- xbar).^2) / (length(x) - 1))
end

function standard_error(x)
    return sample_sd(x) / sqrt(length(x))
end

function rsd_percent(x)
    return 100.0 * sample_sd(x) / mean_value(x)
end

function unknown_concentration(response, slope, intercept)
    return (response - intercept) / slope
end

values = [1.02, 1.05, 0.99]

println("mean=", round(mean_value(values), digits=6))
println("sample_sd=", round(sample_sd(values), digits=6))
println("standard_error=", round(standard_error(values), digits=6))
println("rsd_percent=", round(rsd_percent(values), digits=6))
println("unknown_concentration=", round(unknown_concentration(0.95, 0.30, 0.02), digits=6))
